File: project/src/main/java/br/com/project/service/ReportService.java
package br.com.project.service;

import br.com.project.enums.TransactionCategory;
import br.com.project.exceptions.ResourceNotFoundException;
import br.com.project.models.FinancialTransaction;
import br.com.project.payload.ReportResponse;
import br.com.project.models.Tag;
import br.com.project.models.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.math.BigDecimal;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@Transactional
public class ReportService {

    @Autowired
    private UserService userService;

    @Autowired
    private FinancialTransactionService transactionService;

    public ReportResponse<BigDecimal> totalPayment(Date date, String username) {
        User u = userService.findByUsername(username);

        List<FinancialTransaction> list = transactionService.findByDateAndUser(date, u);

        if(list.isEmpty()) {
            ReportResponse<BigDecimal> reportResponse = new ReportResponse<>();
            reportResponse.setResponseValue(BigDecimal.ZERO);
            return reportResponse;
        }

        ReportResponse<BigDecimal> reportResponse = new ReportResponse<>();
        reportResponse.setResponseValue(
                list.stream()
                        .filter(t -> t.getTransactionCategory().equals(TransactionCategory.PAYMENT))
                        .map(FinancialTransaction::getTransactionValue)
                        .reduce(BigDecimal.ZERO, BigDecimal::add)
        );

        return reportResponse;
    }

    public ReportResponse<BigDecimal> totalReceipt(Date date, String username) {
        User u = userService.findByUsername(username);

        List<FinancialTransaction> list = transactionService.findByDateAndUser(date, u);

        if(list.isEmpty()) {
            ReportResponse<BigDecimal> reportResponse = new ReportResponse<>();
            reportResponse.setResponseValue(BigDecimal.ZERO);
            return reportResponse;
        }

        ReportResponse<BigDecimal> reportResponse = new ReportResponse<>();
        reportResponse.setResponseValue(
                list.stream()
                        .filter(t -> t.getTransactionCategory().equals(TransactionCategory.RECEIPT))
                        .map(FinancialTransaction::getTransactionValue)
                        .reduce(BigDecimal.ZERO, BigDecimal::add)
        );

        return reportResponse;
    }

    public ReportResponse<BigDecimal> accountBalance(Date date, String username) {
        ReportResponse<BigDecimal> totalPayment = totalPayment(date, username);
        ReportResponse<BigDecimal> totalReceipt = totalReceipt(date, username);

        BigDecimal balance = totalReceipt.getResponseValue().subtract(totalPayment.getResponseValue());
        ReportResponse<BigDecimal> responseBalance = new ReportResponse<>();
        responseBalance.setResponseValue(balance);

        return responseBalance;
    }

    public ReportResponse<Map<String, BigDecimal>> reportChart(Date date, String username) {
        User u = userService.findByUsername(username);

        List<Tag> tags = u.getTags();

        if(tags.isEmpty()) {
            throw new ResourceNotFoundException("Você ainda não possui nenhuma tag");
        }

        Map<String, BigDecimal> map = new HashMap<>();

        for(Tag tag: tags) {
            // Verifica se a tag não possui nenhuma transação vinculada a ela, se for o caso
            if(tag.getTransactions().isEmpty()) {
                continue;
            }
            if(
                    tag.getTransactions().stream()
                            .filter(t -> t.getTransactionDate().equals(date))
                            .filter(t -> t.getTransactionCategory().equals(TransactionCategory.PAYMENT))
                            .toList()
                            .isEmpty()
            ) continue;

            map.put(
                    tag.getTagName(),
                    tag.getTransactions().stream()
                            .filter(t -> t.getTransactionDate().equals(date))
                            .filter(t -> t.getTransactionCategory().equals(TransactionCategory.PAYMENT))
                            .map(FinancialTransaction::getTransactionValue)
                            .reduce(BigDecimal.ZERO, BigDecimal::add)
            );
        }

        ReportResponse<Map<String, BigDecimal>> reportResponse = new ReportResponse<>();
        reportResponse.setResponseValue(map);
        return reportResponse;
    }
}

File: src/main/kotlin/br/com/project/projetoIntegrador/service/ReportService.kt
package br.com.project.projetoIntegrador.service

import br.com.project.projetoIntegrador.enums.TransactionCategory
import br.com.project.projetoIntegrador.exceptions.ResourceNotFoundException
import br.com.project.projetoIntegrador.models.User
import br.com.project.projetoIntegrador.payload.ReportPayload
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.LocalDate

@Service
class ReportService(
    @Autowired private val userService: UserService,
    @Autowired private val financialTransactionService: FinancialTransactionService
) {

    fun totalPayment(date: LocalDate, user: User): BigDecimal {
        val list = financialTransactionService.findByDateAndUser(date, user)

        if (list.isEmpty()) return BigDecimal.ZERO

        return list.stream().filter { t -> t.transactionCategory == TransactionCategory.PAYMENT }
            .map { t -> t.transactionValue }
            .reduce(BigDecimal.ZERO, BigDecimal::add)
    }

    fun totalReceipt(date: LocalDate, user: User): BigDecimal {
        val list = financialTransactionService.findByDateAndUser(date, user)

        if (list.isEmpty()) return BigDecimal.ZERO

        return list.stream().filter { t -> t.transactionCategory == TransactionCategory.RECEIPT }
            .map { t -> t.transactionValue }
            .reduce(BigDecimal.ZERO, BigDecimal::add)
    }

//    fun accountBalance(date: LocalDate, username: String): Map<String, BigDecimal> {
//        val totalPayment = totalPayment(date, username)
//        val totalReceipt = totalReceipt(date, username)
//        val balance = (totalReceipt["Recebimentos"] ?: BigDecimal.ZERO)
//            .subtract(totalPayment["Pagamentos"])
//        return mapOf("Saldo" to balance)
//    }

    fun reportChart(date: LocalDate, user: User): Map<String, BigDecimal> {
        val tags = user.tags

        if (tags.isEmpty()) throw ResourceNotFoundException("Você ainda não possui nenhuma tag")

        val map: MutableMap<String, BigDecimal> = mutableMapOf()

        for(tag in tags) {
            if (
                tag.transactions.stream()
                    .filter{ transaction -> transaction.transactionDate.year == date.year &&
                                transaction.transactionDate.month == date.month
                    }
                    .filter{ transaction ->
                        transaction.transactionCategory == TransactionCategory.PAYMENT
                    }
                    .toList().isEmpty()
            ) continue

            val total = tag.transactions.stream()
                .filter { transaction -> transaction.transactionDate.year == date.year &&
                        transaction.transactionDate.month == date.month
                }
                .filter { t -> t.transactionCategory == TransactionCategory.PAYMENT }
                .map { t -> t.transactionValue }
                .reduce(BigDecimal.ZERO, BigDecimal::add)

            map[tag.tagName] = total
        }

        return map
    }

    fun getReport(username: String): ReportPayload? {
        val user = userService.getUser(username)
        val date = LocalDate.now()
        val totalPayment = totalPayment(date, user)
        val totalReceipt = totalReceipt(date, user)
        val accountBalance = totalReceipt.subtract(totalPayment)
        val chartData = reportChart(date, user)
        val map = mapOf<String, BigDecimal>(
            "Saldo" to accountBalance,
            "Receitas" to totalReceipt,
            "Despesas" to totalPayment
        )

        return ReportPayload(map, chartData)
    }
}